### PR TITLE
mgr/nfs: enable reporting err status in Responder decorator, enhance err reporting in nfs export apply

### PR DESCRIPTION
**Pull request by dparmar18**
Prior to this patch, using `ceph nfs export apply -i <nfs_cluster> -i <{json/conf}_file>` would report something like this when the export update/creation failed:

```
$ ceph nfs export apply my-nfs -i exports.json 
[
  {
    "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.1'",
    "state": "error"
  }
]
Error EIO:
```

The last line is incomplete and on top of that `EIO` would be thrown as errno, it makes sense to have a generalised errno since the json/conf file can have multiple export blocks and there can be multiple failures while updating/creating exports but this doesn't make sense when there is a single export block or there is a single export update/creation failure. From my conversation with @adk3798 and @phlogistonjohn in the orch weekly meeting about two weeks back, we decided to:

a) if multiple export blocks?
- in case of multiple failure: report EIO with the err status string
- in case of single failure: report the respective errno with the err status string

b) if single export block?
- report the respective errno with the err status string

This PR makes the changes that were required to be done in the `Responder` class in `pybind/mgr/object_format.py` to make it capable of returning an err status and changes in the `AppliedExportResults` class to adjust the errno and err status string as per the number of failures.

Fixes: https://tracker.ceph.com/issues/62641
Signed-off-by: Dhairya Parmar <dparmar@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by dparmar18** [comment]
could get the errno thing working but the error status is not working, even tried sending a hard coded string but didn't work, not sure why; im investigating it

**Comment by dparmar18** [comment]
the error string should be the third member in this code right
https://github.com/ceph/ceph/blob/63871e38f28e855a1fc4e2e0b82e7d2a4d8a1692/src/pybind/mgr/object_format.py#L533

but it just won't print anything, for the time being let's say my code is buggy yet hard coding it to something like 
```
return retval, body, "error"
```

it would still send this out

```
sh-4.4$ ceph nfs export apply my-nfs -i ex.json 
[
  {
    "msg": "Failed to apply export: export FSAL user_id must be 'nfs.my-nfs.1'",
    "state": "error"
  }
]
Error EINVAL:
```

strange.... will continue debugging tomorrow now

**Comment by phlogistonjohn** [comment]
Hi @dparmar18 I lightly altered your code to try it out (to return a fake error & status) and it worked for me.

```[root@ceph0 ceph]# ceph nfs export apply nfs1  -i - < x1
Inferring fsid 2e7475ac-5244-11ee-99ff-525400220000
Inferring config /var/lib/ceph/2e7475ac-5244-11ee-99ff-525400220000/mon.ceph0/config
Using ceph image with id '9516e6e39ae9' and tag 'dev' created on 2023-09-13 14:36:39 +0000 UTC
quay.io/phlogistonjohn/ceph@sha256:7c065a760639ff3f3d3badb09ec5c53dee99d37e5d2cafff1db7f90e61e596c7
[
  {
    "pseudo": "/fs1",
    "state": "updated"
  }
]
Error EAGAIN: BONK BONK
```

So, while I have a few small comments on the object_format and nfs code changes the main functionality seems to work for me!

**Review by phlogistonjohn — COMMENTED** [review]
Generally the object_format.py changes look pretty good to me. I would suggest making those changes in a seperate commit with updates to the unit tests in src/pybind/mgr/tests/test_object_format.py:

Here, I'll do one for free ;-)
Update the Retty class:
```diff
 class Retty:
-    def __init__(self, v) -> None:
+    def __init__(self, v, status="") -> None:
         self.value = v
+        self.status = status
 
     def mgr_return_value(self) -> int:
         return self.value
 
+    def mgr_status_value(self) -> str:
+        if self.status:
+            return self.status
+        return "NOPE"
+
```
Then add:
```python
@pytest.mark.parametrize(
    "obj, ret",
    [
        ({}, ""),
        ({"fish": "sticks"}, ""),
        (-55, ""),
        (Retty(0), "NOPE"),
        (Retty(-55, "cake"), "cake"),
        (Retty(-50, "pie"), "pie"),
    ],
)
def test_return_status(obj: Any, ret: str):
    rva = object_format.StatusValueAdapter(obj)
    # a ReturnValueAdapter instance meets the ReturnValueProvider protocol.
    assert object_format._is_status_value_provider(rva)
    assert rva.mgr_status_value() == ret

```

**Comment by dparmar18** [comment]
> Generally the object_format.py changes look pretty good to me. I would suggest making those changes in a seperate commit with updates to the unit tests in src/pybind/mgr/tests/test_object_format.py:
> 
> Here, I'll do one for free ;-) Update the Retty class:
> 
> ```diff
>  class Retty:
> -    def __init__(self, v) -> None:
> +    def __init__(self, v, status="") -> None:
>          self.value = v
> +        self.status = status
>  
>      def mgr_return_value(self) -> int:
>          return self.value
>  
> +    def mgr_status_value(self) -> str:
> +        if self.status:
> +            return self.status
> +        return "NOPE"
> +
> ```
> 
> Then add:
> 
> ```python
> @pytest.mark.parametrize(
>     "obj, ret",
>     [
>         ({}, ""),
>         ({"fish": "sticks"}, ""),
>         (-55, ""),
>         (Retty(0), "NOPE"),
>         (Retty(-55, "cake"), "cake"),
>         (Retty(-50, "pie"), "pie"),
>     ],
> )
> def test_return_status(obj: Any, ret: str):
>     rva = object_format.StatusValueAdapter(obj)
>     # a ReturnValueAdapter instance meets the ReturnValueProvider protocol.
>     assert object_format._is_status_value_provider(rva)
>     assert rva.mgr_status_value() == ret
> ```

whoa didn't knew about this, this will be your commit :)

File: src/pybind/mgr/object_format.py [hunk 1]
```diff
@@ -515,6 +548,11 @@ def _return_value(self, obj: Any) -> int:
         """Return a mgr return-value for the given object (usually zero)."""
         return self._retval_provider(obj).mgr_return_value()
 
+    def _return_status(self, obj: Any) -> str:
+        """Return status for the give object (usually empty)"""
```
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/pybind/mgr/object_format.py]: done

File: src/pybind/mgr/object_format.py [hunk 2]
```diff
@@ -366,6 +377,23 @@ def mgr_return_value(self) -> int:
         return self.default_return_value
 
 
+class StatusValueAdapter:
+    """some doc string here"""
```
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/pybind/mgr/object_format.py]: done, added all doc strings

File: src/pybind/mgr/nfs/export.py [hunk 3]
```diff
@@ -167,8 +167,13 @@ class AppliedExportResults:
     def __init__(self) -> None:
         self.changes: List[Dict[str, str]] = []
         self.has_error = False
+        self.exceptions: List[Exception] = []
+        self.err_msg = "some error encountered"
```
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py]: this was just for testing purpose, i was working on a much better solution to send err status I mentioned in the tracker https://tracker.ceph.com/issues/62641.

File: src/pybind/mgr/nfs/export.py [hunk 4]
```diff
@@ -177,7 +182,21 @@ def to_simplified(self) -> List[Dict[str, str]]:
...
+        if self.has_error:
+            if len(self.exceptions) == 1:
+                ex = self.exceptions[0]
+                if isinstance(ex, NFSException):
+                    return ex.errno
+                else:
+                    return -errno.EINVAL
+            if len(self.exceptions) > 1:
+                return -errno.EIO
+        return 0
```
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py]: done, PTAL

**Comment by dparmar18** [comment]
currently writing test cases, will push them soon and mark PR ready for review

**Comment by dparmar18** [comment]
done with writing patch and the test cases, will go through the PR once before marking this as "ready for review"

**Comment by dparmar18** [comment]
tagging @ceph/rgw since this patch will also impact RGW export creation/update using the `ceph nfs export apply -i` interface

**Comment by dparmar18** [comment]
last push was just a rebase

**Review by phlogistonjohn — APPROVED** [review]
LGTM. I can't speak much to the correctness of the teuthology yaml change as I don't know teuthology well enough, but it seems OK to me.  Thanks for working thru this.

**Comment by dparmar18** [comment]
> LGTM. I can't speak much to the correctness of the teuthology yaml change as I don't know teuthology well enough, but it seems OK to me. Thanks for working thru this.

ah that is just for testing this PR out since the current nfs suite fails while running test cases in teuthology, and will be forked out into a separate PR

**Comment by dparmar18** [comment]
last push removed the commit that addressed tracker https://tracker.ceph.com/issues/62870, forked it out into a separate PR https://github.com/ceph/ceph/pull/53594. This PR along with my other NFS PRs will be testable only after https://github.com/ceph/ceph/pull/53594 is merged since right now it test cases fails due complaining 'backend not set'

**Comment by dparmar18** [comment]
took a bit of time to clear issues with test cases where i found out a bug/missing feature that `ceph nfs export create` doesn't allow us to mention export ID of user's choice while this is definitely possible while creating exports using `ceph nfs export apply -i` interface, i created a tracker for it [0]; isn't a big deal to implement but is quite important.

All test cases now pass [1]. This PR is ready for merge. pinging @ceph/cephfs and @ceph/rgw for any suggestions/comments/requests.

[0] https://tracker.ceph.com/issues/63120
[1] http://pulpito.front.sepia.ceph.com/dparmar-2023-10-06_12:13:25-fs:nfs-fix-nfs-apply-err-reporting-distro-default-smithi/

File: src/pybind/mgr/nfs/export.py [hunk 5]
```diff
@@ -167,17 +167,49 @@ class AppliedExportResults:
...
+            if len(self.faulty_export_block_indices) == 1:
+                self.status = f"{str(self.exceptions[0])} for export block" \
+                              f" at index {self.faulty_export_block_indices[0]}"
+            if len(self.faulty_export_block_indices) > 1:
+                indices = ", ".join([str(faulty_block) for faulty_block in
+                                     self.faulty_export_block_indices])
+                self.status = f"{len(self.faulty_export_block_indices)} export" \
+                              f" blocks (at index {indices}) failed to be" \
+                              " created/updated"
+        return self.status
```
Comment (dparmar18) [inline src/pybind/mgr/nfs/export.py:212]: @vshankar the above code is for the err string(which was missing previously) that will be printed along with errno. If you have any suggestions or changes to request do let me know.

Some examples:
```
Export ID changed, Cannot update export for export block at index 1
```


```
2 export blocks (at index 1, 3) failed to be created/updated
```

**Comment by dparmar18** [comment]
i'll test this through fs/nfs, orch/cephadm/smoke-roleless and orch/cephadm/mgr-nfs-upgrade(this one is definitely not necessary but just to be 110% i want to run it through to eliminate chances of ANY issues)

**Comment by dparmar18** [comment]
fs/nfs: http://pulpito.front.sepia.ceph.com/dparmar-2023-10-12_18:11:10-fs:nfs-fix-nfs-apply-err-reporting-distro-default-smithi/

**Comment by dparmar18** [comment]
orch/cephadm/smoke-roleless: http://pulpito.front.sepia.ceph.com/dparmar-2023-10-12_18:15:41-orch:cephadm:smoke-roleless-fix-nfs-apply-err-reporting-distro-default-smithi/

**Comment by dparmar18** [comment]
orch/cephadm/mgr-nfs-upgrade: http://pulpito.front.sepia.ceph.com/dparmar-2023-10-12_18:18:47-orch:cephadm:mgr-nfs-upgrade-fix-nfs-apply-err-reporting-distro-default-smithi/

**Comment by vshankar** [comment]
jenkins test api

**Comment by vshankar** [comment]
@dparmar18 anything that's pending for this PR? I see @phlogistonjohn approval so good to merge given that fs:nfs passes?

**Comment by dparmar18** [comment]
> @dparmar18 anything that's pending for this PR? I see @phlogistonjohn approval so good to merge given that fs:nfs passes?

no blockers as such, just a FYI that err status would be something like this https://github.com/ceph/ceph/pull/53431#discussion_r1356120559. i hope the sentencing is correct or needs any enhancement?

File: src/pybind/mgr/nfs/export.py (hunk 5 above)
Comment (vshankar) [reply src/pybind/mgr/nfs/export.py]: > @vshankar the above code is for the err string(which was missing previously) that will be printed along with errno. If you have any suggestions or changes to request do let me know.
> 
> Some examples:
> 
> ```
> Export ID changed, Cannot update export for export block at index 1
> ```
> 
> ```
> 2 export blocks (at index 1, 3) failed to be created/updated
> ```

Will "index" be useful to the user in this error string?

Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py]: the indices are intended to help figure out which block(s) in the cfg file (conf/json) failed to be created or updated

Comment (vshankar) [reply src/pybind/mgr/nfs/export.py]: Anything other "id"entifier that's a _bit_ more self-descriptive?

Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py]: since this is meant to point out to problematic block(s) in the config file, and considering that the blocks might/might not have expord_id's provided, the only identifier that i can think of could be pseudo path

Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py]: or path, but since the path could anything like `/foo/bar/baz/...`, the err reporting string would look ugly

Comment (vshankar) [reply src/pybind/mgr/nfs/export.py]: Formatting that with json would take care of the ugliness, isn't it? WDYT?

Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py]: this is err status string, we aren't returning json, if json is to be returned it would be something like this  
```
EINVAL:
{
    "paths": foo/bar, bar/baz
}
```

Comment (vshankar) [reply src/pybind/mgr/nfs/export.py]: Why would there be multiple paths? path+pseudo_path?

Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py]: the conf file can have multiple export blocks

Comment (vshankar) [reply src/pybind/mgr/nfs/export.py]: Sure, but then we can have multiple entries in json. Wouldn't that suffice?

Comment (dparmar18) [reply src/pybind/mgr/nfs/export.py]: like how

Comment (vshankar) [reply src/pybind/mgr/nfs/export.py]: (discussion cotd. in slack in the interest of time)

**Comment by vshankar** [comment]
@dparmar18 please update here w.r.t. our discussion yesterday.

**Comment by dparmar18** [comment]
@phlogistonjohn While discussing this PR with @vshankar yesterday; he pointed me to a convention that's present in mgr module for the tuple that is returned as response i.e. `self.errno, "", self.error_str` while in `object_format.py`'s `Responder` class returns `retval, body, ""` now since this PR enables reporting err status it is `retval, body, statusval` but it should be `retval, "", statusval` i.e. the `body` should be part of the `statusval`. 

https://github.com/ceph/ceph/commit/6e234b190841204d076e1b622db543674d0c4b57 introduced this; are there any caveats or edge case due to which this has been kept like this?

**Comment by vshankar** [comment]
@dparmar18 Let's merge this since the tests pass. Please create a tracker for the comment https://github.com/ceph/ceph/pull/53431#issuecomment-1823873370 and hold back the backport till the tracker is taken care of.

**Review by vshankar — APPROVED** [review]

**Comment by dparmar18** [comment]
> @dparmar18 Let's merge this since the tests pass. Please create a tracker for the comment [#53431 (comment)](https://github.com/ceph/ceph/pull/53431#issuecomment-1823873370) and hold back the backport till the tracker is taken care of.

@vshankar here's the tracker https://tracker.ceph.com/issues/63692
